server/aesdsocket.c

```diff
@@ -0,0 +1,316 @@
+		if(strchr(buff,'\n')!=NULL)
+		{
+			break;
+		}
+		
+		mul_factor<<=1;
+		size_t new_buff_size=mul_factor*BUFFER_SIZE;
+		char * new_buff=(char *)realloc(buff,new_buff_size);
+		if(new_buff==NULL)
+		{
+			syslog(LOG_ERR,"realloc failed");
+			free(buff);
+			return -1;
+		}
+		buff=new_buff;
+		buff_size=new_buff_size;
+	}
+	
+	if(write(data_fd,buff,total_rx)!=-1)
+	{
+		fdatasync(data_fd);
+	}
+	else
+	{
+		syslog(LOG_ERR,"writing received data into file failed");
+		free(buff);
+		return -1;
+	}
+	
+	free(buff);
+	return 0;
+	
+}
+
+
+int main(int argc, char **argv)
+{
+    openlog("aesdsocket_log",LOG_PID|LOG_CONS,LOG_USER);
+    bool in_daemon_mode = false;
+    
+    if((argc >=2) && strcmp(argv[1],"-d") == 0)
+    {
+    	in_daemon_mode = true;
+    }
+          
+    socklen_t addr_len;
+    struct addrinfo hints;
+    struct sockaddr_storage addr;
+    int flag,data_fd;
+    int yes=1;
+    
+    memset(&hints,0,sizeof(hints)); //empty the struct
+    hints.ai_flags=AI_PASSIVE;
+    hints.ai_family=AF_INET;
+    hints.ai_socktype=SOCK_STREAM;
+    
+    if((flag=getaddrinfo(NULL,"9000",&hints,&res))!=0)
+    {
+    	syslog(LOG_ERR,"failed to get socket address");
+    	closelog();
+    	exit(1);
+    }
+    
+    if ((sockfd = socket(res->ai_family, res->ai_socktype, res->ai_protocol)) == -1)
+    {
+        syslog(LOG_ERR, "socket() failed");
+        closelog();
+        freeaddrinfo(res);
+        exit(1);
+    }
+    
+     // Set socket options to allow reuse of address
+    if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &(int){1}, sizeof(int)) != 0) {
+        syslog(LOG_ERR, "setsockopt() failed");
+        closelog();
+        freeaddrinfo(res);
+        exit(1);
+    }
+    if (bind(sockfd, res->ai_addr, res->ai_addrlen) == -1) 
+    {
+        syslog(LOG_ERR, "bind() failed");
+        closelog();
+        freeaddrinfo(res);
+        exit(1);
+    }
+    
+    if(in_daemon_mode)
+    {
+    	bool daemon_status = daemon_mode();
+    	if(daemon_status ==false)
+    	{
+    	    syslog(LOG_ERR,"daemon creation failed");
+    	    closelog();
+            freeaddrinfo(res);
+    	    exit(1);
+    	 }
+    }
+    
+    if (listen(sockfd, BACKLOG) == -1)
+    {
+        syslog(LOG_ERR, "Listen failed");
+        closelog();
+        freeaddrinfo(res);
+        exit(1);
+    }
+    
+    data_fd = open("/var/tmp/aesdsocketdata",  O_RDWR |O_CREAT | O_TRUNC,0666);
+    if(data_fd ==-1)
+    {
+    	syslog(LOG_ERR,"/var/tmp/aesdsocketdata file failed to open");
+    	closelog();
+        freeaddrinfo(res);
+    	exit(1);
+    }
+    
+    init_sigaction();
+    addr_len=sizeof(addr);
+    while (!caught_sig)
+    {
+        
+        fd = accept(sockfd, (struct sockaddr *)&addr, &addr_len);
+        if (fd == -1)
+        {
+            syslog(LOG_ERR, "accept() failed: %s", strerror(errno));
+            continue;
+        }
+    	inet_ntop(addr.ss_family, &(((struct sockaddr_in*)&addr)->sin_addr), client_ip, sizeof(client_ip));
```

Comment:
Ensure that the return types are properly checked at all relevant points in the code to avoid any unexpected behavior.